scheduler/core/components/optimizer/greedymax.py

```diff
@@ -374,6 +374,26 @@ def _charge_time(observation: Observation, atom_start: int = 0, atom_end: int =
             observation.sequence[n_atom].observed = True
             observation.sequence[n_atom].qa_state = QAState.PASS
 
+    def _update_score(self, program) -> None:
+        """Update the scores of the incomplete groups in the scheduled program"""
+
+        print("Call score_program")
+        program_calculations = self.selection.score_program(program)
+
+        print("Rescore incomplete schedulable_groups")
+        for unique_group_id in program_calculations.top_level_groups:
+            group_data = program_calculations.group_data_map[unique_group_id]
+            group, group_info = group_data
+            schedulable_group = self.selection.schedulable_groups[unique_group_id]
+            print(f"{unique_group_id} {schedulable_group.group.exec_time()} {schedulable_group.group.total_used()}")
+            print(f"\tOld max score: {np.max(schedulable_group.group_info.scores[0]):7.2f} new max score[0]: "
+                  f"{np.max(group_info.scores[0]):7.2f}")
+            # update scores in schedulable_groups if the group is not completely observed
+            if schedulable_group.group.exec_time() >= schedulable_group.group.total_used():
+                for ii in range(len(group_info.scores)):
+                    schedulable_group.group_info.scores[ii] = group_info.scores[ii]
```

Comment:
It seems like you shouldn't have to do all this work here... we can leave it for now, but I'm going to look at making the rescoring handle this detail.

In any case, it would be much easier to just set the whole `schedulable_groups.group_info.scores` in one fell swoop rather than loop over the elements.

Instead of:

```python
if schedulable_group.group.exec_time() >= schedulable_group.group.total_used():
    for ii in range(len(group_info.scores)):
        schedulable_group.group_info.scores[ii] = group_info.scores[ii]
```

I think this would work:

```python
if schedulable_group.group.exec_time() >= schedulable_group.group.total_used():
    schedulable_group.group_info.scores = group_info.scores
```

And if not that, then this:

```python
if schedulable_group.group.exec_time() >= schedulable_group.group.total_used():
    schedulable_group.group_info.scores = group_info.scores[:]
```

Reply:
When I try one of your options pycharm says "'GroupInfo' object attribute 'scores' is read-only". I tried running it anyway but I get the error "dataclasses.FrozenInstanceError: cannot assign to field 'scores'"

Reply:
but this does work
            if schedulable_group.group.exec_time() >= schedulable_group.group.total_used():
                schedulable_group.group_info.scores[:] = group_info.scores[:]

Reply:
Let's leave it as is for now, and I'll unfreeze the scores... then we can get rid of the loop later.
There's really no rush to do so.

Reply:
Yeah, that will work fine. I'll delay holding off changing the frozen status of scores until we have everything working.